src/devices/machine/cs8900a.cpp

```diff
@@ -0,0 +1,1522 @@
+                    ready,error);
+#endif
+    }
+}
+
+void cs8900a_device::cs8900_set_receiver(int enabled)
+{
+    rx_enabled = enabled;
+    rx_state = CS8900_RX_IDLE;
+
+    rxevent_read_mask = 3; /* was L or H byte read in RXEVENT? */
+}
+
+void cs8900a_device::cs8900_set_transmitter(int enabled)
+{
+    tx_enabled = enabled;
+    tx_state = CS8900_TX_IDLE;
+    
+    cs8900_set_tx_status(0,0);
+}
+
+void cs8900a_device::device_reset(void)
+{
+    // Flush the inbound packet queue
+    std::queue<std::vector<u8>> empty_queue;
+    std::swap(m_frame_queue, empty_queue);
+
+    /* initialize visible IO register and PacketPage registers */
+    memset( cs8900_ioregs, 0, CS8900_COUNT_IO_REGISTER );
+    memset( cs8900_packetpage, 0, MAX_PACKETPAGE_ARRAY );
+
+    /* according to page 19 unless stated otherwise */
+    SET_PP_32(CS8900_PP_ADDR_PRODUCTID,      0x0900630E ); /* p.41: 0E630009 for Rev. D; reversed order! */
+    SET_PP_16(CS8900_PP_ADDR_IOBASE,         0x0300);
+    SET_PP_16(CS8900_PP_ADDR_INTNO,          0x0004); /* xxxx xxxx xxxx x100b */
+    SET_PP_16(CS8900_PP_ADDR_DMA_CHAN,       0x0003); /* xxxx xxxx xxxx xx11b */
+
+    /* according to descriptions of the registers, see definitions of 
+    CS8900_PP_ADDR_CC_... and CS8900_PP_ADDR_SE_... above! */
+
+    SET_PP_16(CS8900_PP_ADDR_CC_RXCFG,       0x0003);
+    SET_PP_16(CS8900_PP_ADDR_CC_RXCTL,       0x0005);
+    SET_PP_16(CS8900_PP_ADDR_CC_TXCFG,       0x0007);
+    SET_PP_16(CS8900_PP_ADDR_CC_TXCMD,       0x0009);
+    SET_PP_16(CS8900_PP_ADDR_CC_BUFCFG,      0x000B);
+    SET_PP_16(CS8900_PP_ADDR_CC_LINECTL,     0x0013);
+    SET_PP_16(CS8900_PP_ADDR_CC_SELFCTL,     0x0015);
+    SET_PP_16(CS8900_PP_ADDR_CC_BUSCTL,      0x0017);
+    SET_PP_16(CS8900_PP_ADDR_CC_TESTCTL,     0x0019);
+
+    SET_PP_16(CS8900_PP_ADDR_SE_ISQ,         0x0000);
+    SET_PP_16(CS8900_PP_ADDR_SE_RXEVENT,     0x0004);
+    SET_PP_16(CS8900_PP_ADDR_SE_TXEVENT,     0x0008);
+    SET_PP_16(CS8900_PP_ADDR_SE_BUFEVENT,    0x000C);
+    SET_PP_16(CS8900_PP_ADDR_SE_RXMISS,      0x0010);
+    SET_PP_16(CS8900_PP_ADDR_SE_TXCOL,       0x0012);
+    SET_PP_16(CS8900_PP_ADDR_SE_LINEST,      0x0014);
+    SET_PP_16(CS8900_PP_ADDR_SE_SELFST,      0x0016);
+    SET_PP_16(CS8900_PP_ADDR_SE_BUSST,       0x0018);
+    SET_PP_16(CS8900_PP_ADDR_SE_TDR,         0x001C);
+    
+    SET_PP_16(CS8900_PP_ADDR_TXCMD,          0x0009);
+
+    /* 4.4.19 Self Status Register, p. 65
+       Important: set INITD (Bit 7) to signal device is ready */
+    SET_PP_16(CS8900_PP_ADDR_SE_SELFST,      0x0896);
+
+    cs8900_recv_control = GET_PP_16(CS8900_PP_ADDR_CC_RXCTL);
+    
+    /* spec: mac address is undefined after reset.
+       real HW: keeps the last set address. */
+    for(int i=0;i<6;i++)
+        SET_PP_8(CS8900_PP_ADDR_MAC_ADDR+i,cs8900_ia_mac[i]);
+    
+    /* reset state */
+    cs8900_set_transmitter(0);
+    cs8900_set_receiver(0);
+}
+
+void cs8900a_device::device_start(void)
```

Comment:
Consider saving state (almost making it visible to the debugger).

Reply:
Gladly-- where can I find an example of doing this?

Reply:
Ah, I think I found examples in other devices.  Are there things that should and should not be saved?

Reply:
The general rule for savestates is variables that are driven by runtime rather than by the normal setup of the device.  So for instance you wouldn't save a ROM area because it's immutable and loaded at startup, but any CPU-writable registers and buffers you definitely would.

Reply:
@rb6502 Can do.  For this, that's the IO registers, the PacketPage, and the class' internal state, so, basically, every member variable.  The inbound frame queue is just a holding spot for frames delivered from the network subsystem so that subsequent reads can spool them off; that likely shouldn't be saved.

Reply:
Right, because the external network won't come up in the same state.